### PR TITLE
aes: refactor NI backend

### PR DESCRIPTION
**Pull request by tarcieri**
- Uses array operations on the block buffer
- Uses loops where they don't appear to impact performance

### CODE REVIEW

**Comment by tarcieri** [comment]
Note: I haven't systematically determined this doesn't change the generated assembly. Preliminary benchmarking shows negligible impact.

File: aes/src/ni/utils.rs [hunk 1]
```diff
@@ -1,5 +1,8 @@
 //! Utility functions
 
+// TODO(tarcieri): check performance impact / generated assembly changes
+#![allow(clippy::needless_range_loop)]
```
Comment (tarcieri) [inline aes/src/ni/utils.rs:4]: Strangely clippy's suggested refactoring seemed to negatively impact performance?(!)

**Review by newpavlov — APPROVED** [review]

File: aes/src/ni/utils.rs (hunk 1 above)
Comment (newpavlov) [reply aes/src/ni/utils.rs:4]: Yeah, I've seen such behavior as well. Iterator based code probably makes it harder for LLVM to unroll the loops. Also subjectively, I think for such loops range loops result in a better, easier to understand code.

File: aes/src/ni/aes128/expand.rs [hunk 2]
```diff
@@ -19,9 +19,9 @@ macro_rules! expand_round {
         t1 = _mm_xor_si128(t1, t3);
         t1 = _mm_xor_si128(t1, t2);
 
-        _mm_store_si128($enc_keys.as_mut_ptr().offset($pos), t1);
+        $enc_keys[$pos] = t1.into();
         let t1 = if $pos != 10 { _mm_aesimc_si128(t1) } else { t1 };
-        _mm_store_si128($dec_keys.as_mut_ptr().offset($pos), t1);
+        $dec_keys[$pos] = t1.into();
```
Comment (newpavlov) [inline aes/src/ni/aes128/expand.rs:24]: Do you need this `into()`? IIUC `t1` has type `__m128i`, same as the array elements.

Comment (tarcieri) [reply aes/src/ni/aes128/expand.rs]: Weird, wonder why clippy didn't catch that...

Anyway, fixed in a802938

**Pull request merged**